pkg/metrics/metrics.go

```diff
@@ -0,0 +1,64 @@
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// MetricsPath is the path that the Metrics are exposed at.
+	MetricsPath = "/metrics"
+
+	// MetricsPort is the port that the Metrics are exposed at.
+	MetricsPort = 8383
+)
+
+var (
+	// Wraps the default RoundTripperFunc with functions that record prometheus metrics.
+	roundTripper promhttp.RoundTripperFunc
+)
+
+// init enables marketplace to serve prometheus metrics.
+func init() {
+	// Register metrics for the operator with the prometheus.
+	log.Info("[metrics] Registering marketplace metrics")
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	log.Info("[metrics] Serving marketplace metrics")
+	http.Handle(MetricsPath, promhttp.Handler())
+	port := fmt.Sprintf(":%d", MetricsPort)
+	go http.ListenAndServe(port, nil)
+}
+
+// GetRoundTripperFunc returns a RoundTripperFunc that Wraps the
+// default RoundTripperFunc with functions that record prometheus
+// metrics.
+func GetRoundTripperFunc() promhttp.RoundTripperFunc {
+	return roundTripper
+}
+
+// registerMetrics registers marketplace prometheus metrics.
+func registerMetrics() error {
+	// appRegistryRequestCounter is a prometheus counter vector that stores
+	// information about reaching out to an AppRegistry when reconciling
+	// OperatorSources.
+	appRegistryRequestCounter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "app_registry_request_count",
+			Help: "A counter that stores the results of reaching out to an AppRegistry.",
+		}, []string{"code", "method", "opsrc"},
+	)
+
+	// Register all of the metrics in the standard registry.
+	prometheus.MustRegister(appRegistryRequestCounter)
+
+	// Wrap the default RoundTripper with middleware.
+	roundTripper = appRegistryRoundTripperCounter(appRegistryRequestCounter, http.DefaultTransport)
```

Comment:
non-blocking: did you look into making a similar wrapper for `request_duration_seconds`?

Reply:
It's pretty easy - I'll add it to this PR.

Reply:
It will look similar to:
```
# HELP app_registry_request_duration_seconds A histogram of AppRegistry request latencies.
# TYPE app_registry_request_duration_seconds histogram
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.005"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.01"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.025"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.05"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.1"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.25"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="0.5"} 0
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="1"} 1
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="2.5"} 1
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="5"} 1
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="10"} 1
app_registry_request_duration_seconds_bucket{opsrc="certified-operators",le="+Inf"} 1
app_registry_request_duration_seconds_sum{opsrc="certified-operators"} 0.951886609
app_registry_request_duration_seconds_count{opsrc="certified-operators"} 1
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.005"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.01"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.025"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.05"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.1"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.25"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="0.5"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="1"} 0
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="2.5"} 1
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="5"} 1
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="10"} 1
app_registry_request_duration_seconds_bucket{opsrc="community-operators",le="+Inf"} 1
app_registry_request_duration_seconds_sum{opsrc="community-operators"} 1.367932254
app_registry_request_duration_seconds_count{opsrc="community-operators"} 1
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.005"} 0
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.01"} 0
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.025"} 0
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.05"} 0
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.1"} 0
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.25"} 0
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="0.5"} 1
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="1"} 1
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="2.5"} 1
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="5"} 1
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="10"} 1
app_registry_request_duration_seconds_bucket{opsrc="redhat-operators",le="+Inf"} 1
app_registry_request_duration_seconds_sum{opsrc="redhat-operators"} 0.433432808
app_registry_request_duration_seconds_count{opsrc="redhat-operators"} 1
# HELP app_registry_request_total A counter that stores the results of reaching out to an AppRegistry.
# TYPE app_registry_request_total counter
app_registry_request_total{code="200",method="get",opsrc="certified-operators"} 1
app_registry_request_total{code="200",method="get",opsrc="community-operators"} 1
app_registry_request_total{code="200",method="get",opsrc="redhat-operators"} 1
```